server/py/services/api/crud/secrets.py

```
@@ -419,6 +423,117 @@ def get_project_secret(
     def is_internal_project_secret_key(self, key: str) -> bool:
         return key.startswith(self.internal_secrets_key_prefix)
 
+    def store_secret_tokens(
+        self,
+        secret_tokens: list[mlrun.common.schemas.SecretToken],
+        authenticated_user_id: str,
+        authenticated_username: str,
+    ) -> mlrun.common.schemas.StoreSecretTokensResponse:
+        """
+        Validate and store offline tokens as Kubernetes secrets.
+
+        :param secret_tokens: List of SecretToken objects to store.
+        :param authenticated_user_id: User ID to validate token ownership.
+        :param authenticated_username: Username used to name Kubernetes secrets.
+        :return: StoreSecretTokensResponse object with created, updated, and skipped tokens.
+        """
+        if not secret_tokens:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "Failed to store secret tokens – no tokens provided"
+            )
+
+        valid_tokens = self._validate_and_decode_offline_tokens(
+            secret_tokens, authenticated_user_id
+        )
+
+        # TODO: move init iguazio_client
+        # iguazio_client = framework.utils.clients.iguazio.v4.Client()
+
+        # TODO: Enable this once it is implemented in the Iguazio client.
+        # iguazio_client.refresh_access_tokens(secret_tokens)
```

this should be before `_validate_and_decode_offline_tokens` since you use `verify_signature=False` there, so first try to use it and only if it works - then decode it.

---

@quaark The whole point was to perform the check on the mlrun side first and fail fast before sending the request to orca. Then, when decoding, we use `verify_signature=False` because the signature will be verified on the orca side later.

---

fixed

---

same - you can now add it in